### DISCUSSION TITLE
Shaping with unscaled glyph

### DISCUSSION THREAD

**Discussion by ubitux** [harfbuzz/harfbuzz#3023] [Q&A] [answered]
Hi,

I'm working with unscaled glyphs from FreeType (not calling `FT_Set_Char_Size()`, `FT_Load_Glyph()` called with `FT_LOAD_NO_SCALE` flag, and not calling `FT_Render_Glyph()`). Since I'm working on the outlines, I'm basically using the native values (font units I assume) to work in my own normalized space. But by doing so, it appears `hb_buffer_get_glyph_positions()` return `0` for everything. Is there a way to make this work with harfbuzz or do I really need to set a scale somewhere?

**Comment by behdad** [comment, marked as answer]
That's definitely possible.

How are you creating your `hb_face_t` and your `hb_font_t`? What HarfBuzz version?

Generally-speaking, `hb_font_t` defaults its scale to the font UPEM in current versions. In older versions it used to default to 0. You can set it to UPEM by fetching that from `hb_face_get_upem()`.

But if you are using `hb-ft.h` to create your `hb_font_t`, things are more complicated. Show me your code. :)

**Reply by ubitux** [reply]
> How are you creating your `hb_face_t` and your `hb_font_t`?

I'm only working with an `hb_font_t` so far, created through `hb_ft_font_create()`.

> What HarfBuzz version?

harfbuzz 2.8.1, freetype 2.10.4

> Generally-speaking, `hb_font_t` defaults its scale to the font UPEM in current versions. In older versions it used to default to 0. You can set it to UPEM by fetching that from `hb_face_get_upem()`.

Mmh, I will check the hb face API, I missed that.

> But if you are using `hb-ft.h` to create your `hb_font_t`, things are more complicated. Show me your code. :)

I'm doing something simple for now:

```c
#include <stdio.h>
#include <hb.h>
#include <hb-ft.h>
#include FT_OUTLINE_H

int main(int ac, char **av)
{
    if (ac != 3) {
        fprintf(stderr, "%s <fontfile> <text>\n", av[0]);
        return 0;
    }

    const char *fontfile = av[1];
    const char *text = av[2];

    FT_Library ft_library;
    FT_Face ft_face;
    if (FT_Init_FreeType(&ft_library))
        return 1;
    if (FT_New_Face(ft_library, fontfile, 0, &ft_face))
        return 1;

#if 0
    if (FT_Set_Char_Size(ft_face, 50 * (1<<6), 0, 0, 0))
        return 1;
#endif

    printf("loaded font family %s\n", ft_face->family_name);
    if (ft_face->style_name)
        printf("* style: %s\n", ft_face->style_name);
    printf("* num glyphs: %ld\n", ft_face->num_glyphs);
    printf("* bbox xmin:%ld xmax:%ld ymin:%ld ymax:%ld\n",
           ft_face->bbox.xMin, ft_face->bbox.xMax,
           ft_face->bbox.yMin, ft_face->bbox.yMax);
    printf("* units_per_EM: %d\n", ft_face->units_per_EM);
    printf("* ascender:  %d\n", ft_face->ascender);
    printf("* descender: %d\n", ft_face->descender);
    printf("* height: %d\n", ft_face->height);
    printf("* max_advance_[width:%d height:%d]\n",
           ft_face->max_advance_width, ft_face->max_advance_height);
    printf("* underline_[position:%d thickness:%d]\n",
           ft_face->underline_position, ft_face->underline_thickness);

    hb_font_t *hb_font = hb_ft_font_create(ft_face, NULL);
    if (!hb_font)
        return 1;

    hb_buffer_t *buffer = hb_buffer_create();
    if (!hb_buffer_allocation_successful(buffer))
        return 1;

    hb_buffer_add_utf8(buffer, text, strlen(text), 0, strlen(text));
    hb_buffer_guess_segment_properties(buffer);
    hb_shape(hb_font, buffer, NULL, 0);

    const int nb_glyphs = hb_buffer_get_length(buffer);
    const hb_glyph_info_t *glyph_info = hb_buffer_get_glyph_infos(buffer, NULL);
    const hb_glyph_position_t *glyph_pos = hb_buffer_get_glyph_positions(buffer, NULL);

    for (int i = 0; i < nb_glyphs; i++) {
        const hb_codepoint_t glyph_id = glyph_info[i].codepoint;
        FT_Load_Glyph(ft_face, glyph_id, FT_LOAD_DEFAULT | FT_LOAD_NO_SCALE);
        //FT_Outline_Decompose(&slot->outline, &outline_funcs, s);

        const hb_glyph_position_t *pos = &glyph_pos[i];
        printf("cp:[%d] advance[x:%d y:%d] offset[x:%d y:%d]\n",
               glyph_id,
               pos->x_advance, pos->y_advance,
               pos->x_offset, pos->y_offset);
    }

    return 0;
}
```

```console
% ./text-test /usr/share/fonts/adobe-source-code-pro/SourceCodePro-Black.otf abc
loaded font family Source Code Pro
* style: Black
* num glyphs: 1568
* bbox xmin:-214 xmax:814 ymin:-454 ymax:1060
* units_per_EM: 1000
* ascender:  984
* descender: -273
* height: 1257
* max_advance_[width:600 height:1257]
* underline_[position:-100 thickness:50]
cp:[28] advance[x:0 y:0] offset[x:0 y:0]
cp:[29] advance[x:0 y:0] offset[x:0 y:0]
cp:[30] advance[x:0 y:0] offset[x:0 y:0]
```

Uncommenting `FT_Set_Char_Size()` gives something for the x-advance.

EDIT: complete code

**Reply by behdad** [reply]
> created through `hb_ft_font_create()`.

Okay, yeah. `hb-ft` was never updated to work with `FT_LOAD_NO_SCALE`. I'll try to look into fixing it, or offer you alternatives, soon.

**Reply by ubitux** [reply]
Thanks! No hurry though, I'm still tinkering with this.

It felt like `NO_SCALE` was what I wanted because I was going to rescale anyway so I wanted the "cleanest" data, but maybe that's not the smartest thing to do. Typically, I might want to work in point units so that the scale is consistent between fonts. What I wanted to avoid at all cost was an arbitrary pixel scaling, more than working on raw font units.

Edit: also note that no specifying the `FT_LOAD_NO_SCALE` flag in the above code doesn't change the behavior (as I'm not calling the render nor a set char size). I'm probably not doing something essential here.

**Reply by ubitux** [reply]
So, I looked a bit at what harfbuzz is doing, and if I want to leave the freetype initialization to harfbuzz, I'd do something:
```c
    hb_blob_t *blob = hb_blob_create_from_file(fontfile);
    hb_face_t *face = hb_face_create(blob, 0);
    hb_font_t *font = hb_font_create(face);
    hb_ft_font_set_funcs(font);
    FT_Face ft_face = hb_ft_font_get_face(font);
```

Looking at the code, it appears that harfbuzz will actually call `FT_Set_Char_Size()`, causing freetype to prepare all kind of pixel-oriented metrics and scaling. It also seems the selected default scale is `1000`.

I guess I'll live with this scaling, but at the same time I will likely not use this method but call `FT_Set_Char_Size()` myself. The reason is a bit stupid, but I just don't like `hb_ft_font_set_funcs()` not having proper error management in case something fails within freetype.

I might also do `FT_Set_Char_Size(ft_face, 16 * 64)` instead of `1000` so that the size is exactly 16pt (in 26.6 format). It might not make any difference though.
